src/Http/DataIteratorBinaryStream.php

```diff
@@ -181,15 +181,17 @@ public function read($length)
         if ($this->eof) {
             return "";
         }
-        $this->rowgen->next();
-        $row = $this->rowgen->current();
         if (!$this->rowgen->valid()) {
             $this->eof = true;
             return chr(0x04);
         }
+        $row = $this->rowgen->current();
```

Comment:
I'm not sure this is right according to the documentation. https://www.php.net/manual/en/iterator.valid.php

```
This method is called after Iterator::rewind() and Iterator::next() to check if the current position is valid. 
```

You're changing it to call valid before calling next.

Reply:
I think you just need to explicitly call rewind in the constructor to fix this problem.

Reply:
@driusan I unfortunately does not seem to work. The [generator documentation](https://www.php.net/manual/en/language.generators.comparison.php) also mentions that _generators are forward-only, meaning that it should be compared to an iterator with a dummy rewind function defined_.

Reply:
Also, the documentation above is for Iterator. The [valid() method for Generators](https://www.php.net/manual/en/generator.valid.php) is used to detect if the end of the loop has been reached.